### ISSUE TITLE
Auto update hashes in nix

### ISSUE THREAD

**Issue by brianmay** [teslamate-org/teslamate#4156]
Purpose of this bug report is to document my current findings on this issue in a public place.

I would like a script that can automatically update the hashes in the flakes.

https://github.com/teslamate-org/teslamate/blob/9b2de5ffa0d0943c9bd89825af681fea14af91df/flake.nix#L27
https://github.com/teslamate-org/teslamate/blob/9b2de5ffa0d0943c9bd89825af681fea14af91df/flake.nix#L35

Here is a solution for npm: https://sitr.us/2024/03/08/nix-npm-and-dependabot.html/

Mix is the problem, I don't know obtain the new hash without building with fakeHash, and then replacing the hash with the hash in the error.

Google search showed me this project which attempted to solve this problem https://github.com/argoproj/argo-workflows/pull/12520 with the following shell script:

```
# this is reliant on the vendorHash being on line 195
sed -i '195s/vendorHash = \"\([^\"]*\)\"/vendorHash = ""/g' ./dev/nix/flake.nix
NIX_HASH=$(nix --log-format raw build ./dev/nix 2>&1 | grep  "got: " | awk '{ split($0,a,"got:    "); print a[2] }')
sed -i '195s|vendorHash = \"\([^\"]*\)\"|vendorHash = "'$NIX_HASH'"|g' ./dev/nix/flake.nix
echo "Changed Nix hash to : $NIX_HASH"
rm -rf result
```

Unfortunately this is broken, it will try to build the project with the old hash, and if the old hash exists, it will be the wrong version.

**Comment by brianmay** [comment]
I asked about this on the matrix channel https://matrix.to/#/!RRerllqmbATpmbJgCn:nixos.org/$HxXrDb80i36hBG3uY4kE8PaX_G3NsM4bKFInux3Q-Xg?via=nixos.org&via=matrix.org&via=tchncs.de

![image](https://github.com/user-attachments/assets/8891d8b2-77b7-446a-9899-04c0e8831071)

Probably need to find a better forum where the maintainers of the elxir and npm code can see it.

**Comment by brianmay** [comment]
Haven't looked into detail,  but this might help (?) https://docs.determinate.systems/guides/automatically-fix-hashes-in-github-actions/

**Comment by brianmay** [comment]
Trying this out on my project. https://github.com/brianmay/phone_db/

Some points:

* Instructions are not 100% correct, had to make some minor changes.

* The Action fails on successfully updating hash, which was confusing at first.

* The npm hash has to be placed in the nix file, not a reference to an external file.

* Wonder if this will work with our split nix system. Actually seems a bit like a magically black box right now.

* Ad of course this is determinate systems...

But when it does work it seems to work really well. So far :-)

**Comment by brianmay** [comment]
The dependabot PR ends up having no associated checks with it, makes me nervous I haven't got this 100% right still.

**Comment by brianmay** [comment]
The one issue I have with this solution: it requires the Determinate nix installer. And I haven't been able to get any nix caching solution working with the Determinate nix installer that doesn't require payment of monthly fees.

**Comment by brianmay** [comment]
I have this working now to the point I am almost happy with it.

Have needed to create a new github user for the actual commits. This is because if you use the github actor to make commits, CI will not run on those commits.

It also makes it easy to check for commit loops. Just check the github.actor value is not the one that makes the commits.

An example https://github.com/brianmay/penguin_memories/blob/main/.github/workflows/update_hashes.yml

The one issue that I still haven't been able to resolve though is getting free caching working. My attempts to get it working in the above file have so far been unsuccessful.

**Comment by JakobLichterfeld** [comment]
What about https://github.com/Mic92/nix-update? Supports `mixFodDeps` and `npmDepsHash`, not sure about cldr

**Comment by brianmay** [comment]
Maybe.

I do see support for `buildNpmPackage` here.

But docs don't mention support for `fetchMixDeps`, which I think we need also.

Unfortunately, no time to look at this right now.

**Comment by JakobLichterfeld** [comment]
> * Ad of course this is determinate systems...

> The one issue I have with this solution: it requires the Determinate nix installer. And I haven't been able to get any nix caching solution working with the Determinate nix installer that doesn't require payment of monthly fees.

Yeah, thats why I want to avoid it after the development with the deprecation of free caching and their lack of communication around.

> What about https://github.com/Mic92/nix-update? Supports `mixFodDeps` and `npmDepsHash`, not sure about cldr

If I find the time, I will try this out

**Comment by brianmay** [comment]
It seems it is possible to use DeterminateSystems/nix-installer-action now with nix-community/cache-nix-action

Which means this works:

https://github.com/brianmay/penguin_memories/blob/main/.github/workflows/update_hashes.yml

Example PR:

https://github.com/brianmay/penguin_memories/pull/786/changes

This one was created with renovate; depedandabot works slightly better (dependabot knows to ignore the added commit - haven't worked that out with renovate yet).

nix-update might be better (?), haven't looked at that yet. I imagine it would follow the same pattern.

**Comment by JakobLichterfeld** [comment]
Perfect. Will take a deeper look when the time is available to me.

**Comment by JakobLichterfeld** [comment]
We cannot use the referenced workflow unchanged because it only runs `determinate-nixd fix hashes` after `nix flake check` has failed with a hash mismatch. Fixed-output derivations are addressed by their declared hash, so an old output already present in the Nix store or restored from the CI cache may be reused without running the fetcher. In that case the build succeeds, no hash mismatch event is produced, and `determinate-nixd` has nothing to fix—even though the cached content no longer matches the updated dependency.

This is especially relevant here because updating `ex_cldr` changes both the Mix dependencies and the required CLDR source, while the old outputs are likely still available in the cache. 

--> I wrote an own helper, which therefore invalidates each pinned hash explicitly before building. CI can invoke that helper and commit its changes automatically, preserving the desired Dependabot automation without depending on whether a stale output happens to be cached. --> PR follow
